app/scripts/lib/rpc-method-middleware/handlers/add-ethereum-chain.js

```diff
@@ -0,0 +1,202 @@
+import { ethErrors } from 'eth-rpc-errors'
+import validUrl from 'valid-url'
+// import { NETWORK_TO_NAME_MAP as DEFAULT_NETWORK_MAP } from '../../../controllers/network/enums'
+import { omit } from 'lodash'
+import { MESSAGE_TYPE } from '../../../../../shared/constants/app'
+import {
+  isPrefixedFormattedHexString,
+  isSafeChainId,
+} from '../../../../../shared/modules/network.utils'
+import { jsonRpcRequest } from '../../../../../shared/modules/rpc.utils'
+
+const addEthereumChain = {
+  methodNames: [MESSAGE_TYPE.ADD_ETHEREUM_CHAIN],
+  implementation: addEthereumChainHandler,
+}
+export default addEthereumChain
+
+async function addEthereumChainHandler(
+  req,
+  res,
+  _next,
+  end,
+  { addCustomRpc, customRpcExistsWith, requestUserApproval },
+) {
+  if (!req.params?.[0] || typeof req.params[0] !== 'object') {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected single, object parameter. Received:\n${req.params}`,
+      }),
+    )
+  }
+
+  const { origin } = req
+
+  const {
+    chainId,
+    chainName = null,
+    blockExplorerUrls = null,
+    nativeCurrency = null,
+    rpcUrls,
+  } = req.params[0]
+
+  const otherKeys = Object.keys(
+    omit(req.params[0], [
+      'chainId',
+      'chainName',
+      'blockExplorerUrls',
+      'iconUrl',
+      'rpcUrls',
+      'nativeCurrency',
+    ]),
+  )
+
+  if (otherKeys.length > 0) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Received unexpected keys on object parameter. Unsupported keys:\n${otherKeys}`,
+      }),
+    )
+  }
+
+  const firstValidRPCUrl = Array.isArray(rpcUrls)
+    ? rpcUrls.find((rpcUrl) => validUrl.isHttpsUri(rpcUrl))
+    : null
+
+  const firstValidBlockExplorerUrl =
+    blockExplorerUrls !== null && Array.isArray(blockExplorerUrls)
+      ? blockExplorerUrls.find((blockExplorerUrl) =>
+          validUrl.isHttpsUri(blockExplorerUrl),
+        )
+      : null
+
+  if (!firstValidRPCUrl) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected an array with at least one valid string HTTPS url 'rpcUrls', Received:\n${rpcUrls}`,
+      }),
+    )
+  }
+
+  if (blockExplorerUrls !== null && !firstValidBlockExplorerUrl) {
+    return end(
+      ethErrors.rpc.invalidParams({
+        message: `Expected null or array with at least one valid string HTTPS URL 'blockExplorerUrl'. Received: ${blockExplorerUrls}`,
+      }),
+    )
+  }
+
+  const _chainId = typeof chainId === 'string' && chainId.toLowerCase()
+
+  if (customRpcExistsWith({ rpcUrl: firstValidRPCUrl, chainId: _chainId })) {
+    return end(
+      ethErrors.rpc.internal({
+        message: `Ethereum chain with the given RPC URL and chain ID already exists.`,
+        data: { rpcUrl: firstValidRPCUrl, chainId },
+      }),
+    )
+  }
```

Comment:
nit: Given that `_chainId` can be `false`, it makes more sense for this check to happen after `isPrefixedFormattedHexString` and `isSafeChainId`.